SkiaSharp blog (#1178)
unoplatform/uno: doc/blog/201907-SkiaSharp-for-wasm-support.md

Inside of the `Uno.SkiaSharp.Views` package you'll find support for the [SKXamlCanvas](https://docs.microsoft.com/en-us/dotnet/api/skiasharp.views.uwp.skxamlcanvas?view=skiasharp-views-uwp-1.68.0) UWP control which enables drawing using Skia in a specific section of the XAML visual tree.

You can experiment with `SKXamlCanvas` by playing with [this sample in our samples repository](https://github.com/nventive/Uno.Samples/tree/master/UI/SkiaSharpTest). Alternatively you can take it for a spin with the following XAML and Code Behind.

XAML:
```xaml
<Grid Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">
    <skia:SKXamlCanvas x:Name="test" PaintSurface="OnPaintSurface" />

> </Grid>

Reviewer: ```suggestion
</Grid>

Code Behind:
```csharp
private void OnPaintSurface(object sender, SKPaintSurfaceEventArgs e)
{
    // the canvas and properties
    var canvas = e.Surface.Canvas;

    // get the screen density for scaling
    var display = DisplayInformation.GetForCurrentView();
    var scale = display.LogicalDpi / 96.0f;
    var scaledSize = new SKSize(e.Info.Width / scale, e.Info.Height / scale);

    // handle the device screen density
    canvas.Scale(scale);

    // make sure the canvas is blank
    canvas.Clear(SKColors.Yellow);

    // draw some text
    var paint = new SKPaint
    {
        Color = SKColors.Black,
        IsAntialias = true,
        Style = SKPaintStyle.Fill,
        TextAlign = SKTextAlign.Center,
        TextSize = 24
    };

    var coord = new SKPoint(
        scaledSize.Width / 2, 
        (scaledSize.Height + paint.TextSize) / 2);
    canvas.DrawText("SkiaSharp", coord, paint);
}
```
```

# SkiaSharp support for WebAssembly

> [Skia](https://skia.org/) is pretty hot lately, and Uno Platform users have been asking for its availability for quite a while, and it is now available through nuget!

Reviewer: Requires :%s/nuget/NuGet/g

![SkiaSharp for WebAssembly](Assets/201906-skiasharp-demo.png)

## Adding SkiaSharp support for WebAssembly

We recently made experiments with CanvasKit, showing that it was possible to run the full SkiaSharp API on top of a custom interop layer, as Mono did not support proper WebAssembly dynamic linking. It was pretty slow, but it worked.

Dynamic linking is now fixed and it enabled the [Windows Calculator's calculation engine to be ported over to WebAssembly](https://platform.uno/a-piece-of-windows-10-is-now-running-on-webassembly-natively-on-ios-and-android/), and it also enables Skia to be compiled to WebAssembly and used through P/Invoke from .NET code.

To enable this, we're using a [fork of the Skia engine](https://github.com/unoplatform/skia) that adds support for both CanvasKit and SkiaSharp's C API, while disabling all the Javascript support. This is required as [double-based APIs are not exportable to JavaScript](https://github.com/emscripten-core/emscripten/commit/ccaf4e74fa9abf51cff8d1d4823f0b4d84bf3eab). This basically makes the Skia module look like a dll, as far as SkiaSharp is concerned.

> Also, Mono supports static linking pretty well, and embedding the Skia engine directly in a Mono AOT-compiled WebAssembly application is available! This makes the [skiasharp-wasm.platform.uno](https://skiasharp-wasm.platform.uno) impressively responsive.

Reviewer: ```suggestion
Thankfully, Mono supports static linking pretty well out-of-the-box and it enabled us to embed the Skia engine directly into [skiasharp-wasm.platform.uno](https://skiasharp-wasm.platform.uno) which is a Mono AOT-compiled WebAssembly application via Uno. Considering how cutting edge this technology is, we are really impressed with how responsive Mono is _right now_.
```

      <skia:SKXamlCanvas x:Name="test" PaintSurface="OnPaintSurface" />
    </Grid>
```

And with some code behind:

```csharp
private void OnPaintSurface(object sender, SKPaintSurfaceEventArgs e)
{
    // the canvas and properties
    var canvas = e.Surface.Canvas;

    // get the screen density for scaling
    var display = DisplayInformation.GetForCurrentView();
    var scale = display.LogicalDpi / 96.0f;
    var scaledSize = new SKSize(e.Info.Width / scale, e.Info.Height / scale);

    // handle the device screen density
    canvas.Scale(scale);

    // make sure the canvas is blank
    canvas.Clear(SKColors.Yellow);

    // draw some text
    var paint = new SKPaint
    {
        Color = SKColors.Black,
        IsAntialias = true,
        Style = SKPaintStyle.Fill,
        TextAlign = SKTextAlign.Center,
        TextSize = 24
    };

    var coord = new SKPoint(
        scaledSize.Width / 2, 
        (scaledSize.Height + paint.TextSize) / 2);

    canvas.DrawText("SkiaSharp", coord, paint);
}
```

You can experiment with [this sample from our samples repository](https://github.com/nventive/Uno.Samples/tree/master/UI/SkiaSharpTest).

## The road ahead

> There are still a few things we need to do for Skia support to be complete. 

Reviewer: ```suggestion
There are a few items outstanding before we will consider support for Skia as complete: 
```

**First**, we needed to fork SkiaSharp to enable WebAssembly support, because of the specificities of the native interop layer, noteable because [`Marshal.GetFunctionPointerForDelegate`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.getfunctionpointerfordelegate?view=netframework-4.8) is not supported yet, and [emscripten's `addFunction`](https://emscripten.org/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html) function pointers feature [needs to be used](https://github.com/unoplatform/Uno.SkiaSharp/blob/uno/binding/SkiaSharp.Wasm/ts/SkiaSharpCanvasKit.ts#L21-L32) for the time being. 

We're [working with the SkiaSharp team](https://github.com/mono/SkiaSharp/issues/876) to add an adjustable interop layer that would enable WebAssembly support to be plugged-in at runtime, allowing proper support for packages such as [`SkiaSharp.Extended`](https://www.nuget.org/packages/SkiaSharp.Extended) to be used without being forked to be supported by WebAssembly.

> **Second**, WebAssembly specifics in terms of function exports require a very specific set of pre-defined methods ([declared here](https://github.com/mono/mono/blob/8d80ccc897c678d7bdae645ca8629b0c5cc0b667/mono/mini/m2n-gen.cs#L30) in mono-wasm). Altering the existing Skia C API signatures to match those "known" methods is possible to some extents, but not for all methods. This is why some methods are still missing from the Skia implementation, and need to be adjusted to use structures intead of parameter, and avoid updating the mono runtime to be supported.

Reviewer 2: ```suggestion
**Second**, WebAssembly specifics in terms of function exports require a very specific set of pre-defined methods ([declared here](https://github.com/mono/mono/blob/8d80ccc897c678d7bdae645ca8629b0c5cc0b667/mono/mini/m2n-gen.cs#L30) in mono-wasm). Altering the existing Skia C API signatures to match those "known" methods is possible to some extent, but not for all methods. This is why some methods are still missing from the Skia implementation, and need to be adjusted to use structures instead of parameters, and avoid updating the Mono runtime to be supported.
```
